core/src/consensus/tendermint/engine.rs

```diff
@@ -165,11 +172,40 @@ impl ConsensusEngine for Tendermint {
             }
             (header.number(), term_id)
         };
+        let rewards = stake::drain_previous_rewards(&mut block.state_mut())?;
+        let client = self
+            .client
+            .read()
+            .as_ref()
+            .ok_or(EngineError::CannotOpenBlock)?
+            .upgrade()
+            .ok_or(EngineError::CannotOpenBlock)?;
+
+        let (start_of_the_next_term, start_of_the_current_term) = {
+            let end_of_the_previous_term = block.state().metadata()?.unwrap().last_term_finished_block_num();
```

Comment:
I'm confusing what is the {current, next, previous} terms.
`term id` is calculated by the header.timestamp. The `block` in this function has different `term` which is calculated from the header, and the block's term is `term id - 1`. Am I right?

What is the current term? Is it from the block's timestamp or from the block?

Reply:
Yes. It's the most confusing part because the current block is the end of the term but the term Id in the metadata is not updated.
The current term in this context is the term that this block closes. The reason which uses the current block to calculate `end_of_the_previous_term` is that because the last block of the previous term is stored in states.